### PR TITLE
Add FindSubtrees

### PR DESCRIPTION
**Pull request by AlCutter**
This PR adds a `FindSubtrees` func which implements the MTC algorithm for determining the one or two subtrees which efficiently cover a range of entries.

### CODE REVIEW

File: proof/proof_test.go [hunk 1]
```diff
@@ -820,3 +820,34 @@ func inclusion(t *testing.T, index, size uint64) Nodes {
...
+
+func TestFindSubtrees(t *testing.T) {
+	for _, tc := range []struct {
+		start, end uint64
+		want       []Subtree
+		wantErr    bool
+	}{
+		{start: 5, end: 5, wantErr: true},
+		{start: 6, end: 5, wantErr: true},
+		{start: 0, end: 1, want: []Subtree{{Start: 0, End: 1}}},
```
Comment (phbnf) [inline proof/proof_test.go:832]: The `FindSubtrees` implementation matches 1:1 the reference implementation, so I've got not doubt that it works.

For completeness though, maybe add:
 - {start: 3, end: 4}: i.e one entry that doesn't start at 0 or at a power of 2
 - {start:4, end: 6}: i.e a perfect subtree, that still returns two subtrees

File: testonly/vectors_test.go [hunk 2]
```diff
@@ -123,3 +125,37 @@ func TestSubtreeConsistencyProofVectors(t *testing.T) {
...
+				if _, err := fmt.Fprintf(h, "[%d, %d)\n", start, end); err != nil {
+					t.Fatalf("fmt.Fprintf: %v", err)
+				}
+			} else {
+				subtrees, err := proof.FindSubtrees(start, end)
+				if err != nil {
+					t.Fatalf("FindSubtrees(%d, %d): %v", start, end, err)
+				}
+				switch l := len(subtrees); l {
+				case 1:
```
Comment (phbnf) [inline testonly/vectors_test.go]: Will this branch ever be executed?

File: proof/proof.go [hunk 3]
```diff
@@ -268,6 +268,36 @@ func reverse(ids []compact.NodeID) {
...
 
+// Subtree represents a valid Merkle subtree [Start, End).
+type Subtree struct {
+	Start, End uint64
+}
+
+// FindSubtrees returns one or two subtrees that efficiently cover [start, end).
+// It corresponds to the "Selecting Two Subtrees" procedure described in Section 4.5.1
+// of draft-ietf-plants-merkle-tree-certs.
+func FindSubtrees(start, end uint64) ([]Subtree, error) {
```
Comment (phbnf) [inline proof/proof.go:288]: This matches the reference implementation 1:1, and so it makes sense.

However, I fail to understand why the specs are written like this, do you?

 - [§4.5.1](https://ietf-plants-wg.github.io/merkle-tree-certs/draft-ietf-plants-merkle-tree-certs.html#section-4.5.1) `The procedure either returns [start, end) as a subtree, or two subtrees, left and right, that satisfy the following properties`, so if `[start, end)` is a perfect subtree, then the proof will only include one subtree.
 - The algorithm below already special cases `len([start, end)) == 1` and already supports returning a single subtree. Why not replace `If end - start is one, return a single subtree, [start, end)` with `if [start, end) is a valid subtree, return a single subtree [start, end)`?

I presume (maybe I'm missing on something?) that all callers of `FindSubtrees` will call `isSubtreeValid` first. Should we provide function that does that? (Alternatively, I'm happy to ask on Slack / push to change the specs... but I'm looking forward to hearing what you think first!)

Comment (pav-kv) [reply proof/proof.go:288]: Drive-by. Looks like it does something related to `Decompose`:

https://github.com/transparency-dev/merkle/blob/600c70387f39a0f17b63fb1748e0222569952d3c/compact/range.go#L248

but rounds up `left` and `right` to a power of 2.

Something like:

```golang
left, right := Decompose(begin, end)
mid := begin + left

left, right = roundPow2(left), roundPow2(right)

return mid - left, mid, mid + right
```

The `Decompose` function splits the range in a way that `left` contains all the subtrees whose left sibling is out-of-bounds from the left side. Nicely, all these subtrees are in a prefix of `[begin, end)`. Same, `right` has all subtrees whose right sibling is out-of-bounds from the right side; and it's a suffix. One of them can be empty/zero (when we are a perfect subtree, or leaning one way).

Because of the siblings property, we can round up and be sure that the left and right subtree still ends/begins at the `mid` point.

Comment (pav-kv) [reply proof/proof.go:288]: Well, it doesn't round up `end`, so I guess it's something slightly different.

File: proof/proof_test.go (hunk 1 above)
Comment (AlCutter) [reply proof/proof_test.go:832]: Why not? Done :)

File: testonly/vectors_test.go (hunk 2 above)
Comment (AlCutter) [reply testonly/vectors_test.go]: Good spot, removed.

File: proof/proof.go (hunk 3 above)
Comment (pav-kv) [reply proof/proof.go:288]: When you/they say "subtree", is it a perfect subtree? If so, I struggle to see how it's always possible to cover a range with 2 perfect subtrees if the right one is not allowed to cross `end`. As an example, `[0, 7)` or `[8,15)`.

So it looks like the 2 subtrees can be non-perfect. Then "efficiently cover [start, end)" means something very specific to this spec. Maybe mention what it means for clarity?

Comment (AlCutter) [reply proof/proof.go:288]: > I presume (maybe I'm missing on something?) that all callers of FindSubtrees will call isSubtreeValid first.

I don't _think_ so -  there's no requirement that `start, end` passed into `FindSubtrees` already describes a valid subtree, or did I miss something?

Comment (AlCutter) [reply proof/proof.go:288]: > Drive-by.

Haha, hi Pavel! :)

Comment (AlCutter) [reply proof/proof.go:288]: > When you/they say "subtree", is it a perfect subtree? If so, I struggle to see how it's always possible to cover a range with 2 perfect subtrees if the right one is not allowed to cross `end`. As an example, `[0, 7)` or `[8,15)`.
> 
> So it looks like the 2 subtrees can be non-perfect. Then "efficiently cover [start, end)" means something very specific to this spec. Maybe mention what it means for clarity?

This is MTC-land stuff, here's the definition it gives for subtrees in that context: https://datatracker.ietf.org/doc/html/draft-ietf-plants-merkle-tree-certs-05#name-definition-of-a-subtree

The 2nd doesn't have to be perfect if that would make it go beyond the range.
FWIW, I did suggest using compact ranges to provide a precise covering, but I think that ship had already sailed.

Comment (pav-kv) [reply proof/proof.go:288]: Thanks Al! That makes sense.

> The 2nd doesn't have to be perfect if that would make it go beyond the range.

Then it just means the second `roundPow2` in my speculation is not needed:

```golang
left, _ := Decompose(begin, end)
mid := begin + left
left = roundPow2(left)
return mid - left, mid, end
```

^ my current best guess what this thing does :D

File: proof/proof.go [hunk 4]
```diff
@@ -268,6 +268,43 @@ func reverse(ids []compact.NodeID) {
...
+//
+// It corresponds to the "Selecting Two Subtrees" procedure described in Section 4.5.1
+// of draft-ietf-plants-merkle-tree-certs.
+//
+// Note that:
+//   - the 2nd subtree, if present, is adjacent to the first, and may not be a perfect subtree.
+//   - the returned subtree(s) fully cover the [start, end) range.
+//   - there are no "extra" entries covered past end, but there may be covered entries prior to start.
+//   - the number of entries covered before start is always less than half the size of the first returned subtree.
+func FindSubtrees(start, end uint64) ([]Subtree, error) {
```
Comment (AlCutter) [inline proof/proof.go:288]: @phbnf @pav-kv Any opinions on returning `(start, mid, end, error)` instead of adding the struct?

File: proof/proof.go (hunk 3 above)
Comment (phbnf) [reply proof/proof.go:288]: Nice seeing you here @pav-kv !

FWIW I had the same intuition and started implemented this...and then eventually stopped when it turned out to be too complex and @AlCutter told me he already had something!

Decompose uses the split between `begin-1` and `end` path. This implementation splits at `begin` and `end-1`. So I think that `mid` won't work either :/

I wish that MTC used more compact ranges... but here we are.

Comment (phbnf) [reply proof/proof.go:288]: > I don't think so - there's no requirement that start, end passed into FindSubtrees already describes a valid subtree, or did I miss something?

Apologies, my comment was not clear. I think that all callers who want to decompose a range will have to wrap the call in something along the lines of:

```
if err := isSubtreeValid(start, end); err != nil {
  return Subtree{start, end}, nil
}
return FindSubtrees(start, end)
```

See the reference test vectors which bypass `FindSubtrees` if the tree is already valid.

If newly append leaves cover [4, 6), since [4, 6) is a valid (and perfect) subtree, there's no need to decompose it, the MTC log server can get a signature over it directly, right?
`FindSubtree(4, 6) = Subtree{4, 5}, Subtree{5, 6}` though, so if we don't wrap the call, the MTC log serve would ask for two subtree signatures, instead of one.

Comment (pav-kv) [reply proof/proof.go:288]: Yeah, the spec behaves weirdly. For example, it decides to decompose `[0,2)` even though it's not needed

https://go.dev/play/p/bci8ykKpLiO

Comment (AlCutter) [reply proof/proof.go:288]: > I think that all callers who want to decompose a range will have to wrap the call in something along the lines of

Ah, I see what you mean now 👍.
Then it seems like this function should just do that too, inline, and fall-back to the two subtree algo? Otherwise this is a bit of a sharp edge.
I've added a commit which does that and adjusted the test + comments, wdyt?

Comment (pav-kv) [reply proof/proof.go:288]: The thing is: it's not only the power-of-2 that's the corner case. If `[begin, end)` is already a "right wing" tree, the algorithm will decompose it, but it's possible not to.

For example, for all trees like `[0,n)`, or all trees in `[8,n) /* n <= 16 */` it returns 2 subtress, while one fits the definition.

https://go.dev/play/p/JtFPP6oUVsV (same code, commented out the panics)

Comment (pav-kv) [reply proof/proof.go:288]: Hm, I see, that's what `isSubtreeValid` checks.

File: proof/proof.go (hunk 4 above)
Comment (phbnf) [reply proof/proof.go:288]: I think the struct is better - otherwise the client would need some extra logic to understand what to do when `mid=end`.

File: proof/proof.go (hunk 3 above)
Comment (phbnf) [reply proof/proof.go:288]: I think the specs needs to be disambiguated, I'll ask folks and see what we can do.

**Review by phbnf — APPROVED** [review]

File: proof/proof.go (hunk 4 above)
Comment (pav-kv) [reply proof/proof.go:288]: Returning `end` seems redundant.

Also, anything breaks with `start == end`? Empty range -> empty decomposition `start == mid == end`.

Maybe in some extreme, the error could also be omitted. I imagine almost any func in this stack is checking the bounds. A common way to replace such checks/errors is a type that carries some invariant:

```golang
// Span is blah.
// Invariant: Begin <= End
type Span struct {
	Begin uint64
	End   uint64
}

func (s Span) check() error {
	if s.Begin > s.End {
		// bark
	}
	return nil
}
```

Then anyone who accepts this param as an input has been promised the invariant and doesn't need to check it again. Only the topmost caller will check it once.

Comment (AlCutter) [reply proof/proof.go:288]: @pav-kv Sorry, was on autopilot and merged when I got back from lunch - didn't mean to ignore your comments!

Comment (AlCutter) [reply proof/proof.go:288]: Yeah, it would be nice to drop the `error` and it's a good point about `end`, and the two do kinda go hand in hand.

@phbnf let's see how you end up using this in the MTC log, perhaps we can circle back and change this if it ends up fitting well there. Maybe the equality checks vs the err check end up being a wash, but worth a look.

Comment (pav-kv) [reply proof/proof.go:288]: @AlCutter np, I am but a driver-by here :D
